lightning/src/sign/mod.rs

```diff
@@ -1983,29 +1983,61 @@ impl KeysManager {
 			.expect("Your RNG is busted");
 		unique_start.input(&child_privkey.private_key[..]);
 
-		let seed = Sha256::from_engine(unique_start).to_byte_array();
+		let unique_seed = Sha256::from_engine(unique_start).to_byte_array();
 
 		let commitment_seed = {
 			let mut sha = Sha256::engine();
-			sha.input(&seed);
+			sha.input(&unique_seed);
 			sha.input(&b"commitment seed"[..]);
 			Sha256::from_engine(sha).to_byte_array()
 		};
 		macro_rules! key_step {
 			($info: expr, $prev_key: expr) => {{
 				let mut sha = Sha256::engine();
-				sha.input(&seed);
+				sha.input(&unique_seed);
 				sha.input(&$prev_key[..]);
 				sha.input(&$info[..]);
 				SecretKey::from_slice(&Sha256::from_engine(sha).to_byte_array())
 					.expect("SHA-256 is busted")
 			}};
 		}
-		let funding_key = key_step!(b"funding key", commitment_seed);
-		let revocation_base_key = key_step!(b"revocation base key", funding_key);
-		let payment_key = key_step!(b"payment key", revocation_base_key);
-		let delayed_payment_base_key = key_step!(b"delayed payment base key", payment_key);
-		let htlc_base_key = key_step!(b"HTLC base key", delayed_payment_base_key);
+
+		let funding_key;
+		let revocation_base_key;
+		let payment_key;
+		let delayed_payment_base_key;
+		let htlc_base_key;
+
+		let channel_keys_derivation_version =
+			channel_keys_derivation_version_from_id(channel_keys_id);
+		if channel_keys_derivation_version < 1 {
+			// In LDK versions prior to 0.1 we used to derive the `payment_key` uniquely on a
+			// per-channel basis, which disallowed users to re-derive them if they lost their
+			// `channel_keys_id`.
+			funding_key = key_step!(b"funding key", commitment_seed);
+			revocation_base_key = key_step!(b"revocation base key", funding_key);
+			payment_key = key_step!(b"payment key", revocation_base_key);
+			delayed_payment_base_key = key_step!(b"delayed payment base key", payment_key);
+			htlc_base_key = key_step!(b"HTLC base key", delayed_payment_base_key);
+		} else {
+			funding_key = key_step!(b"funding key", commitment_seed);
+			revocation_base_key = key_step!(b"revocation base key", funding_key);
+			delayed_payment_base_key = key_step!(b"delayed payment base key", revocation_base_key);
+			htlc_base_key = key_step!(b"HTLC base key", delayed_payment_base_key);
+
+			// Starting with LDK v0.1, we derive `payment_key` directly from our seed, allowing
+			// users to re-derive it even when losing all channel state. This allows them to
+			// recover any non-HTLC-encumbered funds in case of data loss if the counterparty is so
+			// nice to force-closure for them.
+			payment_key = {
+				let mut sha = Sha256::engine();
+				sha.input(&self.seed);
+				sha.input(&b"static payment key"[..]);
+				SecretKey::from_slice(&Sha256::from_engine(sha).to_byte_array())
+					.expect("SHA-256 is busted")
+			};
```

Comment:
question: doesn't this derivation create the same `payment_key` across all the `InMemorySigner`'s derived by this `KeysManager` ?

Reply:
Hmm I'm concerned about a privacy leak here - in case of force close of two channels onchain, these commit tx would both have the same scriptpubkey in the `to_remote` output right ?

I know CLN `hsm_tool` allows people to recover their `to_remote` outputs via a trial-and-error method, we could take a look at how they derive their `payment_key`.

See `hsmtool` `guesstoremote <P2WPKH address> <node id> <tries> "<path/to/hsm_secret>"`

Reply:
That `guesstoremote` method is intended to be used in case the local party loses all channel state, and the remote party force closes the channel - the same scenario addressed in this PR.

Reply:
One more clarification: `<node id>` in that method is the node id of the remote party, not the local party.

Reply:
> Yes, this is true and a good point, although note we currently have that issue with destination/shutdown scripts, too. Though we plan to change this with https://github.com/lightningdevkit/rust-lightning/issues/1139 which I considered to pick up as part / as a follow-up of this PR.

I see ! I had overlooked this code :) I let you determine the best path forward here.

> I guess if we manage to completely isolate the payment_key derivation to a interface method we could consider including the counterparty_node_id, or, even better have the user override that method so that it returns a fresh address from their wallet, which would also get rid of the additional spending step for StaticPaymentOutputs.

Can't the user already decide how `payment_key` is derived in `SignerProvider::derive_channel_signer` ?

We are also constrained by the LN spec if the channel has anchors. But I agree that it would be very cool in the non-anchor case to have the funds land directly in the on-chain wallet from the commitment transaction.

I took a brief look at the API of BDK's `Wallet`, and they don't offer an easy way to ask for raw public keys - just addresses - which makes sense considering that the core of the wallet is an output descriptor.

>  But yeah, maybe you're right and we should try to do it 'the right way' right away.

Your call :)